### PR TITLE
Rename have_optional_dependency

### LINKED ISSUE #3864: Rename `have_optional_dependency` to `import_optional_dependency`
We use `have_optional_dependency()` to import optional dependencies across codebase which is a wrapper over `importlib.import_module()` to throw a helpful error message.

### PR DESCRIPTION
**Pull request by kratman**
# Description

Renames `have_optional_dependency()` to `import_optional_dependency()`

Fixes #3864 

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [x] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [x] All tests pass: `$ python run-tests.py --all` (or `$ nox -s tests`)
- [x] The documentation builds: `$ python run-tests.py --doctest` (or `$ nox -s doctests`)

You can run integration tests, unit tests, and doctests together at once, using `$ python run-tests.py --quick` (or `$ nox -s quick`).

## Further checks:

- [x] Code is commented, particularly in hard-to-understand areas
- [x] Tests added that prove fix is effective or that feature works

### CODE REVIEW

**Review by agriyakhetarpal — APPROVED** [review]
Thanks, @kratman! For next time we should look at removing such functions from the public API – I don't think it is being used by anyone currently and does not provide any useful functionality to users either.

**Comment by kratman** [comment]
> Thanks, @kratman! For next time we should look at removing such functions from the public API – I don't think it is being used by anyone currently and does not provide any useful functionality to users either.

Yeah, python's handling of private methods makes me miss C++

**Review by Saransh-cpp — APPROVED** [review]
Thanks! We should also look into adding `__all__` to pybamm modules; not in this PR though.

**Comment by kratman** [comment]
@arjxn-py, @Saransh-cpp I saw this in `latexify()`
```
        if sympy:
            from pybamm.expression_tree.operations.latexify import Latexify
```
I assume this no longer make sense here either. There used to be an "import_optional_dependency" above it

File: pybamm/util.py [hunk 1]
```diff
@@ -352,25 +352,19 @@ def install_jax(arguments=None):  # pragma: no cover
     )
 
 
-# https://docs.pybamm.org/en/latest/source/user_guide/contributing.html#managing-optional-dependencies-and-their-imports
-def have_optional_dependency(module_name, attribute=None):
+def import_optional_dependency(module_name, attribute=None):
```
Comment (arjxn-py) [inline pybamm/util.py:355]:
```suggestion
# https://docs.pybamm.org/en/latest/source/user_guide/contributing.html#managing-optional-dependencies-and-their-imports
def import_optional_dependency(module_name, attribute=None):
```
Not removing this should be fine

**Review by arjxn-py — COMMENTED** [review]
Looks good to me, thanks

**Comment by Saransh-cpp** [comment]
> I assume this no longer make sense here either. There used to be an "import_optional_dependency" above it

Yes! IIRC we did do some hacky stuff to make sympy optional, maybe this was left behind.

**Comment by agriyakhetarpal** [comment]
> We should also look into adding __all__ to pybamm modules; not in this PR though.

I agree, xref: #2427 for public and private API discussions

**Pull request merged**